Nil pointer SEGV when secret not found.

Credentials secret is not found causes SEGV.

Two problems here:
- The secret pointer needs to be check for `nil`.
- The plan controller should not ensure storage when the MigStorage is not `Ready`.

```
E0913 11:08:28.723115   10091 runtime.go:69] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/home/jortel/go/src/github.com/fusor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:76
/home/jortel/go/src/github.com/fusor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/home/jortel/go/src/github.com/fusor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:522
/usr/local/go/src/runtime/panic.go:513
/usr/local/go/src/runtime/panic.go:82
/usr/local/go/src/runtime/signal_unix.go:390
/home/jortel/go/src/github.com/fusor/mig-controller/pkg/cloudprovider/aws.go:98
/home/jortel/go/src/github.com/fusor/mig-controller/pkg/apis/migration/v1alpha1/migstorage_types.go:165
/home/jortel/go/src/github.com/fusor/mig-controller/pkg/apis/migration/v1alpha1/migstorage_types.go:154
/home/jortel/go/src/github.com/fusor/mig-controller/pkg/controller/migplan/storage.go:164
/home/jortel/go/src/github.com/fusor/mig-controller/pkg/controller/migplan/storage.go:65
/home/jortel/go/src/github.com/fusor/mig-controller/pkg/controller/migplan/migplan_controller.go:228
/home/jortel/go/src/github.com/fusor/mig-controller/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215
/home/jortel/go/src/github.com/fusor/mig-controller/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158
/home/jortel/go/src/github.com/fusor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
/home/jortel/go/src/github.com/fusor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
/home/jortel/go/src/github.com/fusor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/usr/local/go/src/runtime/asm_amd64.s:1333
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x108 pc=0x11bb717]

goroutine 256 [running]:
github.com/fusor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/home/jortel/go/src/github.com/fusor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x108
panic(0x14c8480, 0x2620400)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/fusor/mig-controller/pkg/cloudprovider.(*AWSProvider).UpdateCloudSecret(0xc002d7f780, 0x0, 0xc000a27e00)
	/home/jortel/go/src/github.com/fusor/mig-controller/pkg/cloudprovider/aws.go:98 +0x37
github.com/fusor/mig-controller/pkg/apis/migration/v1alpha1.(*MigStorage).UpdateBSLCloudSecret(0xc001432000, 0x18afe80, 0xc0034195e0, 0xc000a27e00, 0xc0035028a0, 0x149c8c0)
	/home/jortel/go/src/github.com/fusor/mig-controller/pkg/apis/migration/v1alpha1/migstorage_types.go:165 +0xa6
github.com/fusor/mig-controller/pkg/apis/migration/v1alpha1.(*MigStorage).BuildBSLCloudSecret(0xc001432000, 0x18afe80, 0xc0034195e0, 0xc000a27c01, 0x0, 0x0)
	/home/jortel/go/src/github.com/fusor/mig-controller/pkg/apis/migration/v1alpha1/migstorage_types.go:154 +0xcf
github.com/fusor/mig-controller/pkg/controller/migplan.PlanStorage.ensureCloudSecret(0x18afe80, 0xc0034195e0, 0x18af9a0, 0xc0035a1c70, 0xc000408ea0, 0xc001432000, 0x0, 0x0)
	/home/jortel/go/src/github.com/fusor/mig-controller/pkg/controller/migplan/storage.go:164 +0x4c
github.com/fusor/mig-controller/pkg/controller/migplan.ReconcileMigPlan.ensureStorage(0x18af940, 0xc0006ca870, 0xc0002b8620, 0xc000408ea0, 0x0, 0x0)
	/home/jortel/go/src/github.com/fusor/mig-controller/pkg/controller/migplan/storage.go:65 +0x5e9
github.com/fusor/mig-controller/pkg/controller/migplan.(*ReconcileMigPlan).Reconcile(0xc0001e6780, 0xc0006befc0, 0x13, 0xc0004716a0, 0x4, 0x2635f00, 0x0, 0x0, 0x0)
	/home/jortel/go/src/github.com/fusor/mig-controller/pkg/controller/migplan/migplan_controller.go:228 +0x3e7
github.com/fusor/mig-controller/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0000f2320, 0x0)
	/home/jortel/go/src/github.com/fusor/mig-controller/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215 +0x18f
github.com/fusor/mig-controller/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1()
	/home/jortel/go/src/github.com/fusor/mig-controller/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158 +0x36
github.com/fusor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0005b1710)
	/home/jortel/go/src/github.com/fusor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x54
github.com/fusor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0005b1710, 0x3b9aca00, 0x0, 0x1, 0xc0005e77a0)
	/home/jortel/go/src/github.com/fusor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xbe
github.com/fusor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc0005b1710, 0x3b9aca00, 0xc0005e77a0)
	/home/jortel/go/src/github.com/fusor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by github.com/fusor/mig-controller/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start
	/home/jortel/go/src/github.com/fusor/mig-controller/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:157 +0x32a
exit status 2
```

--- a/pkg/apis/migration/v1alpha1/migplan_types.go
+++ b/pkg/apis/migration/v1alpha1/migplan_types.go
@@ -228,7 +228,7 @@ func (r *MigPlan) UpdateRegistrySecret(client k8sclient.Client, storage *MigStor
 		return err
 	}
 	if credSecret == nil {
-		return CredSecretNotFound
+		return errors.New("Credentials secret not found.")
 	}
 	secret.Data = map[string][]byte{
 		"access_key": []byte(credSecret.Data[pvdr.AwsAccessKeyId]),

--- a/pkg/apis/migration/v1alpha1/migstorage_types.go
+++ b/pkg/apis/migration/v1alpha1/migstorage_types.go
@@ -33,9 +33,6 @@ const (
 	GCP   = pvdr.GCP
 )
 
-// Error
-var CredSecretNotFound = errors.New("Credentials secret not found.")
-
 // MigStorageSpec defines the desired state of MigStorage
 type MigStorageSpec struct {
 	BackupStorageProvider  string `json:"backupStorageProvider"`
@@ -161,6 +158,9 @@ func (r *MigStorage) UpdateBSLCloudSecret(client k8sclient.Client, cloudSecret *
 	if err != nil {
 		return err
 	}
+	if secret == nil {
+		return errors.New("Credentials secret not found.")
+	}
 	provider := r.GetBackupStorageProvider()
 	provider.UpdateCloudSecret(secret, cloudSecret)
 	return nil
@@ -185,6 +185,9 @@ func (r *MigStorage) UpdateVSLCloudSecret(client k8sclient.Client, cloudSecret *
 	secret, err := r.GetVolumeSnapshotCredSecret(client)
 	if err != nil {
 		return err
+	}
+	if secret == nil {
+		return errors.New("Credentials secret not found.")
 	}
 	provider := r.GetBackupStorageProvider()
 	provider.UpdateCloudSecret(secret, cloudSecret)

--- a/pkg/controller/migplan/registry.go
+++ b/pkg/controller/migplan/registry.go
@@ -22,7 +22,7 @@ func (r ReconcileMigPlan) ensureMigRegistries(plan *migapi.MigPlan) error {
 		log.Trace(err)
 		return err
 	}
-	if storage == nil {
+	if storage == nil || !storage.Status.IsReady() {
 		return nil
 	}
 	clusters, err := r.planClusters(plan)

--- a/pkg/controller/migplan/storage.go
+++ b/pkg/controller/migplan/storage.go
@@ -22,7 +22,7 @@ func (r ReconcileMigPlan) ensureStorage(plan *migapi.MigPlan) error {
 		log.Trace(err)
 		return err
 	}
-	if storage == nil {
+	if storage == nil || !storage.Status.IsReady() {
 		return nil
 	}
 	clusters, err := r.planClusters(plan)

